### PR TITLE
Fix Issue #330

### PR DESCRIPTION
**Pull request by mtkumar123**
e.g. Fixes #0 

**Checklist**:

- [ ] The GitHub pipeline is OK (green), 
      meaning that both `pylint` (static code analyzer) and `black` (code formatter) are happy with the changes of this PR.

- [x] A unit test is covering the code added / modified by this PR

- [x] This PR is ready to be merged 

- [ ] In case of a new feature, docstrings have been added, with also some documentation in the `docs/` folder

- [ ] A mention of the change is present in `CHANGELOG.md`

- [ ] The PR description or comment contains a picture of a cute animal (not mandatory but encouraged 😉)

### CODE REVIEW

**Comment by Lucas-C** [comment]
Thanks for the PR @mtkumar123!
Lets see if the GitHub Actions pipeline is OK.

In the meantime:

@allcontributors please add @mtkumar123 for code

**Comment by Lucas-C** [comment]
There are a few things `pylint` is unhappy with:
https://github.com/PyFPDF/fpdf2/runs/6107443709?check_suite_focus=true
```
Run pylint fpdf test tutorial/tuto*.py
************* Module fpdf.fpdf
fpdf/fpdf.py:3406:16: W0707: Consider explicitly re-raising using the 'from' keyword (raise-missing-from)
************* Module fpdf.errors
fpdf/errors.py:53:0: C0301: Line too long (193/135) (line-too-long)
************* Module test.errors.test_FPDF_errors
test/errors/test_FPDF_errors.py:24:41: W0612: Unused variable 'e' (unused-variable)
```

File: fpdf/errors.py [hunk 1]
```diff
@@ -30,3 +30,24 @@ def __str__(self):
...
+        self.encoding = encoding
+        self.start = start
+        self.end = end
+        self.reason = reason
+        self.character = character
+
+    def __repr__(self):
+        return (
+            f"{self.__class__.__name__}"
+            f"{repr(self.encoding), repr(self.start), repr(self.end), repr(self.reason)}"
```
Comment (Lucas-C) [inline fpdf/errors.py:49]: `self.character` is missing there

File: fpdf/errors.py [hunk 2]
```diff
@@ -30,3 +30,24 @@ def __str__(self):
...
+
+
+class FPDFUnicodeEncodingException(FPDFException):
+    """Error is thrown when a character that cannot be encoded by the chosen encoder is provided"""
+
+    def __init__(self, encoding, start, end, reason, character):
+        super().__init__()
+        self.encoding = encoding
+        self.start = start
+        self.end = end
```
Comment (Lucas-C) [inline fpdf/errors.py:42]: Both `.end` & `.reason` seem unused: why storing them as attributes?

File: test/errors/test_FPDF_errors.py [hunk 3]
```diff
@@ -19,6 +17,15 @@ def test_add_page_throws_without_page():
     assert str(e.value) == msg
 
 
+def test_encoding_exception():
+    pdf = fpdf.FPDF()
+    pdf.add_page()
+    pdf.set_font("Helvetica", size=15)
+    with pytest.raises(FPDFException) as e:
```
Comment (Lucas-C) [inline test/errors/test_FPDF_errors.py:24]: 👍 for the unit test

You could check that a `FPDFUnicodeEncodingException` is raised here

File: fpdf/errors.py [hunk 4]
```diff
@@ -30,3 +30,24 @@ def __str__(self):
         else:
             res = self.argument
         return f"{self.__class__.__name__ }: {res}"
+
+
+class FPDFUnicodeEncodingException(FPDFException):
```
Comment (Lucas-C) [inline fpdf/errors.py:35]: Good idea to create a subclass!

File: fpdf/errors.py (hunk 1 above)
Comment (mtkumar123) [reply fpdf/errors.py]: Added.

File: fpdf/errors.py (hunk 2 above)
Comment (mtkumar123) [reply fpdf/errors.py]: Removed.

File: test/errors/test_FPDF_errors.py (hunk 3 above)
Comment (mtkumar123) [reply test/errors/test_FPDF_errors.py]: Changed to FPDFPUnicodeEncodingException

File: fpdf/fpdf.py [hunk 5]
```diff
@@ -3408,7 +3400,16 @@ def normalize_text(self, txt):
         # - for TTF unicode fonts: unicode object (utf8 encoding)
         # - for built-in fonts: string instances (encoding: latin-1, cp1252)
         if not self.unifontsubset and self.core_fonts_encoding:
-            return txt.encode(self.core_fonts_encoding).decode("latin-1")
+            try:
+                return txt.encode(self.core_fonts_encoding).decode("latin-1")
+            except UnicodeEncodeError as e:
```
Comment (Lucas-C) [inline fpdf/fpdf.py:3405]: We try to avoid one-letter variable names: could you please rename this as `error`?

**Comment by Lucas-C** [comment]
Thanks for improving things!
I added one extra comment.

Also, there are still a few issues with `pylint` 😢 :
https://github.com/PyFPDF/fpdf2/runs/6113851354?check_suite_focus=true
```
Run pylint fpdf test tutorial/tuto*.py
************* Module fpdf.errors
fpdf/errors.py:51:104: C0303: Trailing whitespace (trailing-whitespace)
fpdf/errors.py:38:40: W0613: Unused argument 'end' (unused-argument)
fpdf/errors.py:38:45: W0613: Unused argument 'reason' (unused-argument)
```

You could try running Pylint on your computer:
```
pip install pylint
pylint
```

**Comment by mtkumar123** [comment]
> Thanks for improving things! I added one extra comment.
> 
> Also, there are still a few issues with `pylint` 😢 : https://github.com/PyFPDF/fpdf2/runs/6113851354?check_suite_focus=true
> 
> ```
> Run pylint fpdf test tutorial/tuto*.py
> ************* Module fpdf.errors
> fpdf/errors.py:51:104: C0303: Trailing whitespace (trailing-whitespace)
> fpdf/errors.py:38:40: W0613: Unused argument 'end' (unused-argument)
> fpdf/errors.py:38:45: W0613: Unused argument 'reason' (unused-argument)
> ```
> 
> You could try running Pylint on your computer:
> 
> ```
> pip install pylint
> pylint
> ```

Sorry, I didn't check pylint. Did it now, and it looks good! Let me know if there is anything else. Thanks!

**Comment by Lucas-C** [comment]
The GitHub Actions pipeline is passing, great! 🥳

Merging this now, thank you very much for your contribution @mtkumar123

**Pull request merged**